### DISCUSSION TITLE
Errors on startup and deploying the repository

### DISCUSSION THREAD

**Discussion by acehog** [lcintron/star-wars-ge-droid-pod#4] [Q&A]
Total noob here and have no idea what I am doing wrong. Does this puTTy screenshot provide any clues?
It stops running anything after the last line. 
First section is running startup after deploying the repository clone.
Second section is deploying the repository clone.
Thanks for any help.

---Error when running startup:---
pi@raspberrypi:~ $ cd /home/pi/star-wars-ge-droid-pod
pi@raspberrypi:~/star-wars-ge-droid-pod $ chmod +x startup.sh
pi@raspberrypi:~/star-wars-ge-droid-pod $ ./startup.sh
  File "/home/pi/star-wars-ge-droid-pod/ui/main.py", line 58
    except requests.exceptions.RequestException as e:
TabError: inconsistent use of tabs and spaces in indentation
pi@raspberrypi:~/star-wars-ge-droid-pod $

---Errors deployng the cloned repository:---
pi@raspberrypi:~ $ git clone https://github.com/lcintron/star-wars-ge-droid-pod.git
Cloning into 'star-wars-ge-droid-pod'...
remote: Enumerating objects: 345, done.
remote: Total 345 (delta 0), reused 0 (delta 0), pack-reused 345
Receiving objects: 100% (345/345), 2.46 MiB | 897.00 KiB/s, done.
Resolving deltas: 100% (202/202), done.
pi@raspberrypi:~ $ cd /home/pi/star-wars-ge-droid-pod/server
pi@raspberrypi:~/star-wars-ge-droid-pod/server $ npm install

> usb@1.6.0 install /home/pi/star-wars-ge-droid-pod/server/node_modules/usb
> prebuild-install --verbose || node-gyp rebuild

internal/modules/cjs/loader.js:550
    throw err;
    ^

Error: Cannot find module 'bl'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:548:15)
    at Function.Module._load (internal/modules/cjs/loader.js:475:25)
    at Module.require (internal/modules/cjs/loader.js:598:17)
    at require (internal/modules/cjs/helpers.js:11:18)
    at Object.<anonymous> (/home/pi/star-wars-ge-droid-pod/server/node_modules/tar-stream/extract.js:2:10)
    at Module._compile (internal/modules/cjs/loader.js:654:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:665:10)
    at Module.load (internal/modules/cjs/loader.js:566:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:506:12)
    at Function.Module._load (internal/modules/cjs/loader.js:498:3)
make: Entering directory '/home/pi/star-wars-ge-droid-pod/server/node_modules/usb/build'
  CC(target) Release/obj.target/libusb/libusb/libusb/core.o
  CC(target) Release/obj.target/libusb/libusb/libusb/descriptor.o
  CC(target) Release/obj.target/libusb/libusb/libusb/hotplug.o
  CC(target) Release/obj.target/libusb/libusb/libusb/io.o
  CC(target) Release/obj.target/libusb/libusb/libusb/strerror.o
  CC(target) Release/obj.target/libusb/libusb/libusb/sync.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/poll_posix.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/threads_posix.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/linux_usbfs.o
  CC(target) Release/obj.target/libusb/libusb/libusb/os/linux_udev.o
../libusb/libusb/os/linux_udev.c:40:10: fatal error: libudev.h: No such file or directory
   40 | #include <libudev.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [libusb.target.mk:132: Release/obj.target/libusb/libusb/libusb/os/linux_udev.o] Error 1
make: Leaving directory '/home/pi/star-wars-ge-droid-pod/server/node_modules/usb/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:180:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:12)
gyp ERR! System Linux 5.15.61+
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/pi/star-wars-ge-droid-pod/server/node_modules/usb
gyp ERR! node -v v9.11.2
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok

> bluetooth-hci-socket@0.5.1 install /home/pi/star-wars-ge-droid-pod/server/node_modules/bluetooth-hci-socket
> node-gyp rebuild

make: Entering directory '/home/pi/star-wars-ge-droid-pod/server/node_modules/bluetooth-hci-socket/build'
  CXX(target) Release/obj.target/binding/src/BluetoothHciSocket.o
In file included from /home/pi/.node-gyp/9.11.2/include/node/node.h:63,
                 from /home/pi/.node-gyp/9.11.2/include/node/node_buffer.h:25,
                 from ../src/BluetoothHciSocket.cpp:7:
/home/pi/.node-gyp/9.11.2/include/node/v8.h: In static member function ‘static void v8::V8::RemoveGCPrologueCallback(v8::GCCallback)’:
/home/pi/.node-gyp/9.11.2/include/node/v8.h:10378:7: warning: cast between incompatible function types from ‘v8::GCCallback’ {aka ‘void (*)(v8::GCType, v8::GCCallbackFlags)’} to ‘v8::Isolate::GCCallback’ {aka ‘void (*)(v8::Isolate*, v8::GCType, v8::GCCallbackFlags)’} [-Wcast-function-type]
10378 |       reinterpret_cast<Isolate::GCCallback>(callback));
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pi/.node-gyp/9.11.2/include/node/v8.h: In static member function ‘static void v8::V8::RemoveGCEpilogueCallback(v8::GCCallback)’:
/home/pi/.node-gyp/9.11.2/include/node/v8.h:10385:7: warning: cast between incompatible function types from ‘v8::GCCallback’ {aka ‘void (*)(v8::GCType, v8::GCCallbackFlags)’} to ‘v8::Isolate::GCCallback’ {aka ‘void (*)(v8::Isolate*, v8::GCType, v8::GCCallbackFlags)’} [-Wcast-function-type]
10385 |       reinterpret_cast<Isolate::GCCallback>(callback));
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ../src/BluetoothHciSocket.cpp:8:
../../nan/nan.h: In function ‘void Nan::AsyncQueueWorker(Nan::AsyncWorker*)’:
../../nan/nan.h:2298:7: warning: cast between incompatible function types from ‘void (*)(uv_work_t*)’ {aka ‘void (*)(uv_work_s*)’} to ‘uv_after_work_cb’ {aka ‘void (*)(uv_work_s*, int)’} [-Wcast-function-type]
 2298 |     , reinterpret_cast<uv_after_work_cb>(AsyncExecuteComplete)
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/BluetoothHciSocket.cpp: In member function ‘void BluetoothHciSocket::poll()’:
../src/BluetoothHciSocket.cpp:251:95: warning: ‘v8::Local<v8::Value> Nan::MakeCallback(v8::Local<v8::Object>, v8::Local<v8::String>, int, v8::Local<v8::Value>*)’ is deprecated [-Wdeprecated-declarations]
  251 | ::New<Object>(this->This), Nan::New("emit").ToLocalChecked(), 2, argv);
      |                                                                      ^

In file included from ../src/BluetoothHciSocket.cpp:8:
../../nan/nan.h:1045:46: note: declared here
 1045 |   NAN_DEPRECATED inline v8::Local<v8::Value> MakeCallback(
      |                                              ^~~~~~~~~~~~
../src/BluetoothHciSocket.cpp: In member function ‘void BluetoothHciSocket::emitErrnoError()’:
../src/BluetoothHciSocket.cpp:275:72: warning: ‘v8::Local<v8::Object> v8::Function::NewInstance(int, v8::Local<v8::Value>*) const’ is deprecated: Use maybe version [-Wdeprecated-declarations]
  275 | Local<Value> error = errorConstructor->NewInstance(1, constructorArgs);
      |                                                                      ^

In file included from /home/pi/.node-gyp/9.11.2/include/node/v8.h:26,
                 from /home/pi/.node-gyp/9.11.2/include/node/node.h:63,
                 from /home/pi/.node-gyp/9.11.2/include/node/node_buffer.h:25,
                 from ../src/BluetoothHciSocket.cpp:7:
/home/pi/.node-gyp/9.11.2/include/node/v8.h:3894:31: note: declared here
 3894 |                 Local<Object> NewInstance(int argc, Local<Value> argv[]) const);
      |                               ^~~~~~~~~~~
/home/pi/.node-gyp/9.11.2/include/node/v8config.h:318:3: note: in definition of macro ‘V8_DEPRECATED’
  318 |   declarator __attribute__((deprecated(message)))
      |   ^~~~~~~~~~
../src/BluetoothHciSocket.cpp:282:93: warning: ‘v8::Local<v8::Value> Nan::MakeCallback(v8::Local<v8::Object>, v8::Local<v8::String>, int, v8::Local<v8::Value>*)’ is deprecated [-Wdeprecated-declarations]
  282 | ::New<Object>(this->This), Nan::New("emit").ToLocalChecked(), 2, argv);
      |                                                                      ^

In file included from ../src/BluetoothHciSocket.cpp:8:
../../nan/nan.h:1045:46: note: declared here
 1045 |   NAN_DEPRECATED inline v8::Local<v8::Value> MakeCallback(
      |                                              ^~~~~~~~~~~~
In file included from /home/pi/.node-gyp/9.11.2/include/node/node_buffer.h:25,
                 from ../src/BluetoothHciSocket.cpp:7:
../src/BluetoothHciSocket.cpp: At global scope:
/home/pi/.node-gyp/9.11.2/include/node/node.h:505:7: warning: cast between incompatible function types from ‘void (*)(Nan::ADDON_REGISTER_FUNCTION_ARGS_TYPE)’ {aka ‘void (*)(v8::Local<v8::Object>)’} to ‘node::addon_register_func’ {aka ‘void (*)(v8::Local<v8::Object>, v8::Local<v8::Value>, void*)’} [-Wcast-function-type]
  505 |       (node::addon_register_func) (regfunc),                          \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pi/.node-gyp/9.11.2/include/node/node.h:539:3: note: in expansion of macro ‘NODE_MODULE_X’
  539 |   NODE_MODULE_X(modname, regfunc, NULL, 0)  // NOLINT (readability/null_usage)
      |   ^~~~~~~~~~~~~
../src/BluetoothHciSocket.cpp:499:1: note: in expansion of macro ‘NODE_MODULE’
  499 | NODE_MODULE(binding, BluetoothHciSocket::Init);
      | ^~~~~~~~~~~
In file included from /home/pi/.node-gyp/9.11.2/include/node/node.h:63,
                 from /home/pi/.node-gyp/9.11.2/include/node/node_buffer.h:25,
                 from ../src/BluetoothHciSocket.cpp:7:
/home/pi/.node-gyp/9.11.2/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = node::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)]’:
/home/pi/.node-gyp/9.11.2/include/node/node_object_wrap.h:85:78:   required from here
/home/pi/.node-gyp/9.11.2/include/node/v8.h:9338:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<node::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<node::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
 9338 |                reinterpret_cast<Callback>(callback), type);
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/pi/.node-gyp/9.11.2/include/node/v8.h: In instantiation of ‘void v8::PersistentBase<T>::SetWeak(P*, typename v8::WeakCallbackInfo<P>::Callback, v8::WeakCallbackType) [with P = Nan::ObjectWrap; T = v8::Object; typename v8::WeakCallbackInfo<P>::Callback = void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)]’:
../../nan/nan_object_wrap.h:65:61:   required from here
/home/pi/.node-gyp/9.11.2/include/node/v8.h:9338:16: warning: cast between incompatible function types from ‘v8::WeakCallbackInfo<Nan::ObjectWrap>::Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<Nan::ObjectWrap>&)’} to ‘Callback’ {aka ‘void (*)(const v8::WeakCallbackInfo<void>&)’} [-Wcast-function-type]
  SOLINK_MODULE(target) Release/obj.target/binding.node
  COPY Release/binding.node
make: Leaving directory '/home/pi/star-wars-ge-droid-pod/server/node_modules/bluetooth-hci-socket/build'
npm WARN ge-beacons@0.0.1 No repository field.
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: xpc-connection@0.1.4 (node_modules/xpc-connection):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for xpc-connection@0.1.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"arm"})
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: usb@1.6.0 (node_modules/usb):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: usb@1.6.0 install: `prebuild-install --verbose || node-gyp rebuild`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1

added 58 packages in 158.764s
pi@raspberrypi:~/star-wars-ge-droid-pod/server $

**Comment by mkanoap** [comment]
"TabError: inconsistent use of tabs and spaces in indentation" is python complaining that most of the time things are indented with spaces, but a couple of lines have tabs instead.   The version of python he used when first creating this might not of cared about this, but newer ones do.

It's also complaining that you don't have a library called "BL" installed, there may be a problem with your npm installation.

Are you using python 2.x or python3?  I had to make some small code changes to make it work with python 3, besides fixing the inconsistent use of tabs and spaces.

**Comment by randyandrade** [comment]
*bump* 

Has anyone been able to successfully install and build this with recent software versions?

I've been trying to get this sorted, but I'm running into the same types of things with dependencies not being met, and conflicts between packages. Package names have changed as well, so that doesn't help. Driver version is newer, too.
